simulator/src/main.rs

```diff
@@ -74,18 +95,29 @@ async fn main() {
 
     info!(options = ?*OPTIONS);
 
-    let tcn_broadcast_buffer_size = OPTIONS.num_users * 20;
-    let (tx, _) = tokio::sync::broadcast::channel(tcn_broadcast_buffer_size);
+    //let tcn_broadcast_buffer_size = OPTIONS.num_users * 20;
+    //let (tx, _) = tokio::sync::broadcast::channel(tcn_broadcast_buffer_size);
 
+    let shard_choices = Uniform::new(0u64, OPTIONS.num_shards);
+    let mut channels: HashMap<ShardId, broadcast::Sender<TemporaryContactNumber>> = HashMap::new();
+    for shardid in 0u64..OPTIONS.num_shards {
+	let (tx, _) = tokio::sync::broadcast::channel(OPTIONS.num_users * 20);
+	channels.insert(shardid, tx);
+    }
+    
     let mut users = futures::stream::FuturesUnordered::new();
 
     use std::time::Duration;
     use tokio::time::delay_for;
-
+    
     for id in 0..OPTIONS.num_users {
         // Stagger the start of each user.
         delay_for(Duration::from_millis(1)).await;
-        users.push(tokio::spawn(User::default().run(id, tx.clone())));
+	let shard_id = shard_choices.sample(&mut thread_rng());
+	println!("{}", shard_id);
+	let tx = channels.get(&shard_id).unwrap().clone();
+	// TODO: Do we need to handle None here, as its a one-to-one mapping?
```

Comment:
Which `None` do you mean?  If you mean the `None` that can come from `channels.get()`, you could instead write it using `expect`, which lets you write the invariant you expect to hold as a static string that will be put into a panic message if the invariant is violated:
```
let tx = channels.get(&shard_id)
    .expect("entry must be present because we sample from known keys")
    .clone();
```